Remove UseJsonFallbackSerializer

ReubenBond opened the pull request:
Follow-on from #2193.

Json.NET is not a suitable fallback serializer. It does not handle the complex types which it needs to handle.

sergeybykov commented on src/Orleans/Serialization/SerializationManager.cs:
```diff
@@ -143,36 +138,26 @@ public static object GetUninitializedObjectWithFormatterServices(Type type)
         {
             RegisterBuiltInSerializers();
-
-#if NETSTANDARD
```
Don't we still need this #if and the warning on NETSTANDARD?

ReubenBond replied:
Eep, this entire block should be removed. Thanks for pointing this out!

ReubenBond replied:
Oh, I misread the diff. I'm not sure that we need a warning which the user cannot do anything about. On the other hand, in `GetFallbackSerializer()`, I'm throwing unconditionally on `NETSTANDARD`, which is clearly untenable. What's the other option, though, other than to wait for #2162 for `NETSTANDARD` support.

We could return a fallback serializer which always throws.

sergeybykov replied:
Maybe we should leave the warning here, and let `GetFallbackSerializer()` throw right after?

ReubenBond replied:
No problem, change incoming.

sergeybykov commented on src/Orleans/Serialization/OrleansJsonSerializer.cs:
```diff
@@ -7,16 +7,20 @@
-    internal class OrleansJsonSerializer : IExternalSerializer
+    using Orleans.Providers;
+
+    public class OrleansJsonSerializer : IExternalSerializer
```
Why do we need to make it public?

ReubenBond replied:
Users ought to be able to specify external serializers to use, even if only in certain cases.
This class could be used for storage providers, for example. Currently they use bits of this class (even if they were using it via `SerializationManager`).

sergeybykov commented on src/Orleans/Serialization/SerializationManager.cs:
```diff
@@ -2210,40 +2186,5 @@ public SerializerMethods(DeepCopier deepCopy, Serializer serialize, Deserializer
             public Deserializer Deserialize { get; private set; }
         }
-
-        public static JsonSerializerSettings GetDefaultJsonSerializerSettings()
```
Don't we still need this for other usages of `OrleansJsonSerializer`, e.g. by storage providers?

ReubenBond replied:
See the other comment about moving the method. If you feel it's better to keep the methods on SerializationManager, then I can move them back. It seems messy, though.

A cleaner approach would be for those storage providers to consume an `IExternalSerializer` for all of their serialization, rather than each implementing support for JSON + Orleans serialization.

sergeybykov commented on test/TesterInternal/Serialization/BuiltInSerializerTests.cs:
```diff
@@ -65,8 +59,7 @@ public BuiltInSerializerTests(ITestOutputHelper output)
         }
 
         [Theory, TestCategory("Functional"), TestCategory("Serialization")]
-        [InlineData(SerializerToUse.Default)]
```
What's this change for?

ReubenBond replied:
It makes it easier to add & test other serializers (eg, `IlBasedSerializer`) without modifying every method.

In the (legitimate?) case where one serializer is intended to fail and another is intended to succeed, then we can use `InlineData` to specify that.

I'm happy to revert the change and simply remove the do-nothing `Fallback` option.

sergeybykov commented on src/Orleans/Serialization/OrleansJsonSerializer.cs:
```diff
@@ -7,16 +7,20 @@
         /// </summary>
         /// <returns></returns>
-        internal static JsonSerializerSettings GetDefaultSerializerSettings()
+        public static JsonSerializerSettings GetDefaultSerializerSettings()
```
Same question here.

ReubenBond replied:
This method was essentially moved from `SerializationManager`, but it's used by external providers in order to get serialization settings which are suitable for serializing types which we use (grain refs, etc).

sergeybykov replied:
I don't disagree with your reasoning. But doesn't this mean that all storage providers that use `OrleansJsonSerializer` would have to be updated? If so, then I think it is better to make such a change explicit - as a separate PR, and probably targeting 1.4.0, since we are so close to releasing 1.3.0.

ReubenBond replied:
Agreed

ReubenBond commented:
> It appears as if there's more than just removing the option of using JSON as a fallback serializer here. But maybe I'm just confused.

You're right, I'll split the other changes into a separate PR. They were changes I made while I was cleaning up the JSON fallback serializer option, but they're not as targeted as the title suggests.

ReubenBond commented:
Updated - it's much slimmer now.
I retained the removal of the redundant `[InlineData(Fallback)]` attribute in `BuiltInSerializerTests`, since it previously had the same behavior as `JsonFallback`, which is now removed.
